File: lc_soln/0003_longest_substring_without_repeating_characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int result = 0;
        int start = 0;
        std::unordered_map<char, int> counter;
        for (int i = 0; i < s.size(); ++i) {
            while (counter[s[i]] > 0) {
                counter[s[start]]--;
                start++;
            }
            counter[s[i]]++;
            result = std::max(result, i - start + 1);
        }
        return result;
    }
};

int main() {
    return 0;
}

File: lc_soln/0003_longest_substring_without_repeating_characters/solution.rs
use std::collections::HashMap;

struct Solution;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut result = 0;
        let mut start = 0;
        let mut counter = HashMap::<char, i32>::new();
        let chars: Vec<_> = s.chars().collect();
        for (i, &c) in chars.iter().enumerate() {
            while counter.get(&c).unwrap_or(&0) > &0 {
                *counter.get_mut(&chars[start]).unwrap() -= 1;
                start += 1;
            }
            *counter.entry(c).or_insert(0) += 1;
            result = result.max((i - start + 1) as i32);
        }
        result
    }
}

fn main() {
    let s = "abcabcbb".to_string();
    println!("{}", Solution::length_of_longest_substring(s));
}
